streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -303,18 +303,42 @@ abstract class ZSink[-R, +E, -I, +L, +Z] private (
     } yield push)
 
   /**
-   * Returns the sink that executes this one and times its execution.
+   * Returns a sink that summarizes an effect by computing some value before and
+   * after execution, and then combining the values to produce a summary,
+   * together with the result of the sink.
    */
-  final def timed: ZSink[R with Clock, E, I, L, (Z, Duration)] =
+  final def summarized[R1 <: R, E1 >: E, B, C](
+    summary: ZIO[R1, E1, B]
+  )(f: (B, B) => C): ZSink[R1, E1, I, L, (Z, C)] =
     ZSink {
-      self.push.zipWith(clock.nanoTime.toManaged_) { (push, start) =>
+      self.push.zipWith(summary.either.toManaged_) { (push, start) =>
         push(_).catchAll {
-          case (Left(e), leftover)  => Push.fail(e, leftover)
-          case (Right(z), leftover) => clock.nanoTime.flatMap(stop => Push.emit(z -> (stop - start).nanos, leftover))
+          case (Left(e), leftover) =>
+            Push.fail(start.fold(identity, _ => e), leftover)
+
+          case (Right(z), leftover) =>
+            start match {
+              case Left(e) => Push.fail(e, leftover)
+              case Right(start) =>
+                summary.foldM(
+                  e => Push.fail(e, leftover),
+                  end => Push.emit((z, f(start, end)), leftover)
+                )
+            }
         }
       }
     }
 
+  /**
+   * Returns the sink that executes this one and times its execution.
+   */
+  final def timed: ZSink[R with Clock, E, I, L, (Z, Duration)] =
+    summarized(
+      clock.nanoTime
+    )((start, stop) => (start, stop)).map { case (z, (start, stop)) =>
```

Comment:
Can you calculate the difference in `f` so you don't need the `map`?

Reply:
You can probably just copy the implementation from the `series/2.x` branch now.

Reply:
Done.